yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr

```diff
@@ -0,0 +1,364 @@
+                    3,
+                    |private_context, public_context, slot| {
+                        Set::new(private_context, public_context, slot, ValueNoteMethods)
+                    },
+                ),
+                total_supply: PublicState::new(
+                    private_context,
+                    public_context,
+                    4,
+                    FieldSerialisationMethods,
+                ),
+                pending_shields: Set::new(private_context, public_context, 5, TransparentNoteMethods),
+                public_balances: Map::new(
+                    private_context,
+                    public_context,
+                    6,
+                    |private_context, public_context, slot| {
+                        PublicState::new(
+                            private_context,
+                            public_context,
+                            slot,
+                            FieldSerialisationMethods,
+                        )
+                    },
+                ),
+            }
+        }
+    }
+
+    #[aztec(private)]
+    fn constructor() {
+        // Currently not possible to execute public calls from constructor as code not yet available to sequencer.
+        // let selector = compute_selector("_initialize((Field))");
+        // let _callStackItem = context.call_public_function(context.this_address(), selector, [context.msg_sender()]);
+    }
+
+    #[aztec(public)]
+    fn set_admin(
+        new_admin: AztecAddress,
+    ) {
+        let storage = Storage::init(Option::none(), Option::some(&mut context));
+        assert(storage.admin.read() == context.msg_sender(), "caller is not admin");
+        storage.admin.write(new_admin.address);
+    }
+
+    #[aztec(public)]
+    fn set_minter(
+        minter: AztecAddress,
+        approve: Field,
+    ) {
+        assert((approve == 1) | (approve == 0), "not providing boolean");
+        let storage = Storage::init(Option::none(), Option::some(&mut context));
+        assert(storage.admin.read() == context.msg_sender(), "caller is not admin");
+        storage.minters.at(minter.address).write(approve as Field);
+    }
+
+    #[aztec(public)]
+    fn mint_pub(
+        to: AztecAddress,
+        amount: Field,
+    ) -> Field {
+        let storage = Storage::init(Option::none(), Option::some(&mut context));
+        assert(storage.minters.at(context.msg_sender()).read() == 1, "caller is not minter");
+        let amount = SafeU120::new(amount);
+        let new_balance = SafeU120::new(storage.public_balances.at(to.address).read()).add(amount);
+        let supply = SafeU120::new(storage.total_supply.read()).add(amount);
+
+        storage.public_balances.at(to.address).write(new_balance.value as Field);
+        storage.total_supply.write(supply.value as Field);
+        1
+    }
+
+    #[aztec(public)]
+    fn mint_priv(
+        amount: Field,
+        secret_hash: Field,
+    ) -> Field {
```

Comment:
I understand that the only thing that prevents us from having a private `mint` method is being able to access the list of minters from private-land, and for that we'd need the slow-updates tree, right?

Reply:
Yes, but even if in private I would want it to do a public call to update the `total_supply` such that it is up to date. But yes you could mint directly into a specific user if we got slow-updates tree.